File: mobile/ios/OmenIOS/OmenIOS/App/Api/PlayerSearch.swift
import Foundation

/// `GET /api/players/search` → `players-search.v1`.
///
/// The Trade destination shipped on 2026-08-30 with a bare text field: the user typed a name
/// freehand and pressed Add. **This route already existed** — documented in `api-routes.md` as
/// "Free Trade Analyzer autocomplete", public Sleeper data, no auth, max 10 rows — and native
/// simply never called it. The founder found it in the first minutes of real use:
/// *"players' names don't pop up… it's like the page wasn't wired."* He was right.
///
/// It is the same defect class as `F-SCR-01`: a capability the backend already serves, that the
/// native client did not consume.
struct PlayerSearchResult: Decodable, Equatable, Identifiable {
    let id: String
    let name: String
    let position: String?
    let team: String?
    /// Present only when the backend found no exact/substring result and is
    /// offering a correction. A fuzzy row is never silently committed.
    let matchType: String?

    init(
        id: String,
        name: String,
        position: String?,
        team: String?,
        matchType: String? = nil
    ) {
        self.id = id
        self.name = name
        self.position = position
        self.team = team
        self.matchType = matchType
    }

    enum CodingKeys: String, CodingKey {
        case id, name, position, team
        case matchType = "match_type"
    }

    var isFuzzySuggestion: Bool { matchType == "fuzzy" }

    /// "WR · MIN" — omitted entirely when the provider gives neither, rather than rendering
    /// a stray separator.
    var subtitle: String? {
        let parts = [position, team].compactMap { $0?.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }
        return parts.isEmpty ? nil : parts.joined(separator: " · ")
    }
}

protocol PlayerSearchRepository {
    func search(query: String) async -> Result<[PlayerSearchResult], OmenApiError>
}

struct ApiPlayerSearchRepository: PlayerSearchRepository {
    private let client: OmenApiClient

    init(client: OmenApiClient) {
        self.client = client
    }

    func search(query: String) async -> Result<[PlayerSearchResult], OmenApiError> {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count >= 2 else { return .success([]) }

        // Public route — no bearer. Same posture as `/api/trade/compare`. The query goes
        // through `URLComponents`, never string interpolation into the path.
        return await client.get(
            "api/players/search",
            optionalAccessToken: nil,
            query: ["q": trimmed],
            as: [PlayerSearchResult].self
        )
    }
}

struct StubPlayerSearchRepository: PlayerSearchRepository {
    let result: Result<[PlayerSearchResult], OmenApiError>

    func search(query: String) async -> Result<[PlayerSearchResult], OmenApiError> { result }
}

File: mobile/android/app/src/main/kotlin/com/slopssaloon/omen/app/feature/api/PlayerSearch.kt
package com.slopssaloon.omen.app.feature.api

import org.json.JSONArray

/**
 * `GET /api/players/search` -> `players-search.v1`. iOS mirror: `App/Api/PlayerSearch.swift`.
 *
 * The Trade destination shipped on both platforms with a bare text field: the user typed a name
 * freehand and pressed Add. **This route already existed** — documented in `api-routes.md` as
 * "Free Trade Analyzer autocomplete", public Sleeper data, no auth, max 10 rows — and neither
 * native client ever called it. The founder found it in the first minutes of real use:
 * *"players' names don't pop up... it's like the page wasn't wired."* He was right.
 *
 * Same defect class as `F-SCR-01`: a capability the backend already serves that the native
 * clients do not consume.
 */
data class PlayerSearchResult(
    val id: String,
    val name: String,
    val position: String?,
    val team: String?,
    /** Present only for a correction candidate, never an automatic identity resolution. */
    val matchType: String? = null,
) {
    val isFuzzySuggestion: Boolean get() = matchType == "fuzzy"
    /**
     * "WR - MIN", omitted entirely when the provider gives neither rather than rendering a
     * stray separator against an empty half.
     */
    val subtitle: String?
        get() {
            val parts = listOfNotNull(position, team).map { it.trim() }.filter { it.isNotEmpty() }
            return if (parts.isEmpty()) null else parts.joinToString(" \u00B7 ")
        }

    companion object {
        /**
         * The payload is a bare array. A row missing `id` or `name` is skipped rather than
         * failing the whole list — one malformed player must not blank the picker.
         */
        fun parseList(json: String): List<PlayerSearchResult>? = runCatching {
            val array = JSONArray(json)
            (0 until array.length()).mapNotNull { index ->
                val row = array.optJSONObject(index) ?: return@mapNotNull null
                val id = row.optString("id").takeIf { it.isNotEmpty() } ?: return@mapNotNull null
                val name = row.optString("name").takeIf { it.isNotEmpty() } ?: return@mapNotNull null
                PlayerSearchResult(
                    id = id,
                    name = name,
                    position = row.optString("position").takeIf { it.isNotEmpty() },
                    team = row.optString("team").takeIf { it.isNotEmpty() },
                    matchType = row.optString("match_type").takeIf { it.isNotEmpty() },
                )
            }
        }.getOrNull()
    }
}

interface PlayerSearchRepository {
    suspend fun search(query: String): OmenApiResult<List<PlayerSearchResult>>
}

class ApiPlayerSearchRepository(private val client: OmenApiClient) : PlayerSearchRepository {
    override suspend fun search(query: String): OmenApiResult<List<PlayerSearchResult>> {
        val trimmed = query.trim()
        // Below the minimum the route would 400. Answering locally keeps a one-character
        // keystroke from spending one of the 30-per-minute-per-IP budget.
        if (trimmed.length < MIN_QUERY_LENGTH) return OmenApiResult.Success(emptyList())

        // Public route — no bearer, same posture as `POST /api/trade/compare`.
        return client.getOptionalAuth(
            path = "api/players/search",
            accessToken = null,
            query = mapOf("q" to trimmed),
            decode = PlayerSearchResult::parseList,
        )
    }

    companion object {
        const val MIN_QUERY_LENGTH = 2
    }
}

class StubPlayerSearchRepository(
    private val result: OmenApiResult<List<PlayerSearchResult>>,
) : PlayerSearchRepository {
    override suspend fun search(query: String): OmenApiResult<List<PlayerSearchResult>> = result
}
